quinn/src/endpoint.rs

```diff
@@ -795,3 +758,140 @@ impl std::ops::Deref for EndpointRef {
         &self.0
     }
 }
+
+#[derive(Debug)]
+struct PollState {
+    inner: proto::Endpoint,
+    transmit_state: TransmitState,
+    incoming: VecDeque<proto::Incoming>,
+    connections: ConnectionSet,
+    recv_buf: Box<[u8]>,
+    recv_limiter: WorkLimiter,
+    metas: [RecvMeta; BATCH_SIZE],
+}
+
+impl PollState {
+    fn new(
+        sender: mpsc::UnboundedSender<(ConnectionHandle, EndpointEvent)>,
+        inner: proto::Endpoint,
+        max_receive_segments: usize,
+    ) -> Self {
+        let recv_buf = vec![
+            0;
+            inner.config().get_max_udp_payload_size().min(64 * 1024) as usize
+                * max_receive_segments
+                * BATCH_SIZE
+        ];
+        Self {
+            connections: ConnectionSet {
+                senders: FxHashMap::default(),
+                sender,
+                close: None,
+            },
+            inner,
+            transmit_state: TransmitState::default(),
+            incoming: VecDeque::new(),
+            recv_buf: recv_buf.into(),
+            recv_limiter: WorkLimiter::new(RECV_TIME_BOUND),
+            metas: [RecvMeta::default(); BATCH_SIZE],
+        }
+    }
+    fn poll_socket(
+        &mut self,
+        socket: &Arc<dyn AsyncUdpSocket>,
```

Comment:
We don't need to clone the `Arc`, so we should deref eagerly for clarity:
```suggestion
        socket: &dyn AsyncUdpSocket,
```